http/sys_monitor.go

```diff
@@ -0,0 +1,63 @@
+package http
+
+import (
+	"fmt"
+	"net/http"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/monitor"
+	"github.com/hashicorp/vault/vault"
+)
+
+func handleSysMonitor(core *vault.Core) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Remove the default 90 second timeout so clients can stream indefinitely.
+		// Using core.activeContext means that this will get cancelled properly
+		// when the server shuts down.
+		c, _ := core.GetContext()
+		r = r.Clone(c)
+
+		ll := r.URL.Query().Get("log_level")
+		if ll == "" {
+			ll = "INFO"
+		}
+		logLevel := log.LevelFromString(ll)
+
+		if logLevel == log.NoLevel {
+			respondError(w, http.StatusBadRequest, fmt.Errorf("unknown log level"))
+			return
+		}
+
+		flusher, ok := w.(http.Flusher)
+		if !ok {
+			respondError(w, http.StatusBadRequest, fmt.Errorf("streaming not supported"))
+			return
+		}
+
+		isJson := core.LogFormat() == "json"
+		monitor, _ := monitor.NewMonitor(512, core.Logger(), &log.LoggerOptions{
+			Level:      logLevel,
+			JSONFormat: isJson,
+		})
+
+		logCh := monitor.Start()
+		w.WriteHeader(http.StatusOK)
+
+		// 0 byte write is needed before the Flush call so that if we are using
+		// a gzip stream it will go ahead and write out the HTTP response header
+		w.Write([]byte(""))
+		flusher.Flush()
+
+		// Stream logs until the connection is closed.
+		for {
+			select {
+			case <-r.Context().Done():
+				monitor.Stop()
+				return
+			case log := <-logCh:
+				fmt.Fprint(w, string(log))
```

Comment:
Will do

Reply:
Yeah, you're right, which kind of confuses me.  The `<-r.Context().Done()` only fires when the server itself shuts down, not when the client connection is closed.  The docs for `Request.Context` say:

> For outgoing client requests, the context controls cancellation.

> For incoming server requests, the context is canceled when the client's connection closes, the request is canceled (with HTTP/2), or when the ServeHTTP method returns.

When I read that, I thought we were dealing with an incoming server request here, since this is a handler function, and the context would be canceled when the client connection closed, but in practice, that doesn't happen.

Even when I check the error returned from `fmt.Fprint`, that doesn't get triggered when the client closes its connection either.

What's the correct way to make sure this function exits when the client connection is closed?

Reply:
Thinking on this some more, I wonder if it's a mistake to remove the timeout from the request context.  My thought in doing so was to enable the case where people can type `vault monitor` in a terminal, and leave it as long as they like, and it will always work, even if the server has been idle for some extended period.  Maybe that's not a use case that's worth supporting though?  I'm guessing the timeout on the request context is the server's way of making sure connections get dropped if they're idle for an extended period of time.  It seems like the way it is right now, this handler is leaking connections.

I searched around trying to find the answer to this, but all of it seems to come back to the request context.